Switch assertWarnsOnceRegex logic to check any instead of all.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56424 AutoDispatchBelowAutograd takes no arguments.
* #56423 s/AutoNonVariableTypeMode/AutoDispatchBelowAutograd/
* #56422 Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning.
* **#56434 Switch assertWarnsOnceRegex logic to check any instead of all.**

If we hit multiple TORCH_WARN from different sources when running the
statement, it makes more sense to me that we want to check the regex is
met in any one of the warning messages instead of all messages.

Differential Revision: [D27871946](https://our.internmc.facebook.com/intern/diff/D27871946)

Commit:
Switch assertWarnsOnceRegex logic to check any instead of all.

If we hit multiple TORCH_WARN from different sources when running the
statement, it makes more sense to me that we want to check the regex is
met in any one of the warning messages instead of all messages.

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1469,10 +1469,10 @@ def assertWarnsOnceRegex(self, category, regex=''):
-                                f'{pattern}, {w.message}')
+            self.assertTrue(any([type(w.message) is category for w in ws]))
+            self.assertTrue(
+                any([re.match(pattern, str(w.message)) for w in ws]),
```
I really like the motivation for this change. 

The second argument to assertTrue is a string message that appears to a developer when the assertion fails, right? In this case it'd be something like "foo string, warning message warning message warning message" if there were multiple warning messages? 

While warning messages don't need to be perfect, what do you think about filtering ws using the `type(w.message) is category` check above? That way the warning messages will be from the set that should be matched against, and the previous assertion could just validate this filtered set is nonempty.

mruberry reviewed (approved):
Hey @ailzhang! Great idea.

I'm approving this to unblock you if it's urgent. Your fix looks correct but see my suggestion for a possible tweak that might produce nicer messages when the assertion fails. 

The ASAN failure is in this PR's base, btw, and is not related to this change.

Commit:
Update on "Switch assertWarnsOnceRegex logic to check any instead of all."


If we hit multiple TORCH_WARN from different sources when running the
statement, it makes more sense to me that we want to check the regex is
met in any one of the warning messages instead of all messages.

Differential Revision: [D27871946](https://our.internmc.facebook.com/intern/diff/D27871946)

[ghstack-poisoned]
